### PR TITLE
PWX-34097: Pool delete tests

### PR DESCRIPTION
**Pull request by dgoel-px**
**What this PR does / why we need it**:
Migrates PoolDeleteFunctionalityTest and PoolDeleteNegativeTest from px-test

**Which issue(s) this PR fixes** (optional)
Closes #PWX-34097

**Special notes for your reviewer**:
GKE https://jenkins.pwx.dev.purestorage.com/job/Users/job/Divyam/job/gcp-dmthin/4/
AKS https://jenkins.pwx.dev.purestorage.com/job/Users/job/Divyam/job/aks-dmthin/6/
vSphere https://jenkins.pwx.dev.purestorage.com/job/Users/job/Divyam/job/dmthin-clouddrive-2/14/

### CODE REVIEW

File: tests/basic/storage_pool_test.go [hunk 1]
```diff
@@ -9993,3 +9994,283 @@ func exitPoolMaintenance(poolId string) {
...
+		Migrated from px-test: PoolDeleteFunctionality
+			1. Delete pools till total availble pool is 1.
+			2. Verify after each delete if the right pool got deleted.
+			3. Verify Alerts.
+			4. Try to delete the last pool in the node and it should fail.
+			5. Randomly pick a drive which got freeed up because of pool delete and add it back.
+	*/
+
+	JustBeforeEach(func() {
+		StartTorpedoTest("PoolDelete", "Initiate pool deletion", nil, 0)
```
Comment (chiachunlin) [inline tests/basic/storage_pool_test.go:10041]: "PoolDelete" -> "PoolDeleteFunctionality" ?

File: tests/basic/storage_pool_test.go [hunk 2]
```diff
@@ -9993,3 +9994,283 @@ func exitPoolMaintenance(poolId string) {
...
+		selectedNode := node.GetStorageNodes()[0]
+		nodePools := selectedNode.StoragePools
+
+		poolToAddBack := nodePools[rand.Intn(len(nodePools))]
+		for _, pool := range nodePools {
+			poolID := strconv.Itoa(int(pool.ID))
+			deletePoolAndValidate(selectedNode, poolID)
+		}
+
+		// verify all pools deleted
```
Comment (chiachunlin) [inline tests/basic/storage_pool_test.go:10056]: from your comment in L10033, shouldn't it have 1 pool left?

File: tests/basic/storage_pool_test.go [hunk 3]
```diff
@@ -9993,3 +9994,283 @@ func exitPoolMaintenance(poolId string) {
...
+		dash.VerifyFatal(len(pools) == 0, true, "verify all pools deleted")
+
+		contexts = make([]*scheduler.Context, 0)
+		for i := 0; i < Inst().GlobalScaleFactor; i++ {
+			contexts = append(contexts, ScheduleApplications(fmt.Sprintf("pooldeletefunc-%d", i))...)
+		}
+		ValidateApplications(contexts)
+		defer appsValidateAndDestroy(contexts)
+
+		stepLog = fmt.Sprintf("Adding cloud drive to node [%v] with size [%v]", selectedNode.Name, poolToAddBack.TotalSize/units.GiB)
```
Comment (chiachunlin) [inline tests/basic/storage_pool_test.go:10068]: could you print each step in the test?  which step is it in comment L10033~10037?

File: tests/basic/storage_pool_test.go [hunk 4]
```diff
@@ -9993,3 +9994,283 @@ func exitPoolMaintenance(poolId string) {
...
+			log.FailOnError(err, "volume drive down on node %s", selectedNode.Name)
+
+			poolsAfr, err := Inst().V.ListStoragePools(metav1.LabelSelector{})
+			log.FailOnError(err, "Failed to list storage pools")
+			dash.VerifyFatal(len(poolsAfr) == 1, true, "verify new pool is created")
+			newPoolsMap, err := Inst().V.GetPoolDrives(&selectedNode)
+			log.FailOnError(err, "error getting pool drive from the node [%s]", selectedNode.Name)
+			dash.VerifyFatal(len(newPoolsMap) == 1, true, "verify new drive is created")
+		})
+
```
Comment (chiachunlin) [inline tests/basic/storage_pool_test.go:10101]: is L10035: `Verify Alerts.` verified somewhere?

File: tests/basic/storage_pool_test.go [hunk 5]
```diff
@@ -9993,3 +9994,283 @@ func exitPoolMaintenance(poolId string) {
...
+		stepLog = fmt.Sprintf("Expand newly added pool on node [%s]", selectedNode.Name)
```
Comment (chiachunlin) [inline tests/basic/storage_pool_test.go:10102]: same comment as above. maybe you need to update the test comments? (L10032)

File: tests/basic/storage_pool_test.go [hunk 6]
```diff
@@ -9993,3 +9994,283 @@ func exitPoolMaintenance(poolId string) {
...
+		log.FailOnError(err, "failed to find node for repl add")
+		dash.VerifyFatal(replAddNode != nil, true, "expect to find a node to do repl add")
+		replAddPool := replAddNode.StoragePools[0]
+		// TODO add check that pool doesn't contain metadata
+
+		stepLog = fmt.Sprintf("Increasing the volume repls for %s", contextAppKey)
+		Step(stepLog, func() {
+			log.InfoD(stepLog)
+			log.Infof("Increase repl for volume: %v", testVolume.ID)
+			pxctlCmdFull := fmt.Sprintf("v ha-update -r 3 --node %v %v", replAddPool.Uuid, testVolume.ID)
```
Comment (chiachunlin) [inline tests/basic/storage_pool_test.go:10244]: how do you confirm that the volume to be ha-updated has repl=2?

File: tests/common.go [hunk 7]
```diff
@@ -7814,48 +7821,65 @@ func DeleteGivenPoolInNode(stNode node.Node, poolIDToDelete string, retry bool)
...
-		if exitErr = Inst().V.WaitDriverUpOnNode(stNode, 5*time.Minute); exitErr != nil {
-			log.Errorf("error waiting for driver up after exiting pool maintenance in the node [%v]. Err: %v", stNode.Name, exitErr)
-			return
-		}
-		//Adding wait as even PX is up it is taking some time for pool status to update
-		//when all pools are deleted
-		time.Sleep(1 * time.Minute)
-		cmd := "pxctl sv pool show"
-		var out string
+	// //Wait till the Node goes down
```
Comment (chiachunlin) [inline tests/common.go:7825]: can remove the comment I think?

**Comment by chiachunlin** [comment]
can you add the test results?

File: tests/basic/storage_pool_test.go (hunk 4 above)
Comment (dgoel-px) [reply tests/basic/storage_pool_test.go:10101]: No removed that

File: tests/basic/storage_pool_test.go (hunk 6 above)
Comment (dgoel-px) [reply tests/basic/storage_pool_test.go:10244]: We don't care about that for this test. The volume enters resync is all we are about.

File: drivers/scheduler/k8s/specs/fio-storagepool/fio-config-map.yaml [hunk 8]
```diff
@@ -0,0 +1,21 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: fio-job-config
+data:
+    fio.job: |
+        [global]
+        name=fio-write
```
Comment (lsrinivas-pure) [inline drivers/scheduler/k8s/specs/fio-storagepool/fio-config-map.yaml:8]: any specific why we are creating new spec?

File: tests/basic/storage_pool_test.go [hunk 9]
```diff
@@ -10005,3 +10006,341 @@ func exitPoolMaintenance(poolId string) {
...
+	dash.VerifyFatal(testNode != nil, true, "select test node")
+	return testNode
+}
+
+var _ = Describe("{PoolDeleteFunctionality}", func() {
+	/*
+		Migrated from px-test: PoolDeleteFunctionality
+			1. Delete pools till total availble pool is 1 and verify for each delete that the right pool was deleted.
+			2. Delete the last pool and verify node transitions to storageless.
+			3. Randomly pick a drive which got freeed up because of pool delete and add it back.
```
Comment (lsrinivas-pure) [inline tests/basic/storage_pool_test.go:10084]: nit: freed

File: tests/basic/storage_pool_test.go [hunk 10]
```diff
@@ -10005,3 +10006,341 @@ func exitPoolMaintenance(poolId string) {
...
+			// no need to exit pool maintenance mode
+			err = EnterPoolMaintenance(selectedNode)
+			log.FailOnError(err, "")
+			poolID := strconv.Itoa(int(nodePools[len(nodePools)-1].ID))
+			err = Inst().V.DeletePool(selectedNode, poolID, true)
+			log.FailOnError(err, "")
+
+			poolsMap, err := Inst().V.GetPoolDrives(&selectedNode)
+			log.FailOnError(err, "error getting pool drive from the node [%s]", selectedNode.Name)
+			dash.VerifyFatal(len(poolsMap) == 0, true, "verify all pools deleted")
```
Comment (lsrinivas-pure) [inline tests/basic/storage_pool_test.go:10119]: better if you do refresh node registry and nodes cache will still have this as storagenode and pool count will differ and futher tests referring them might fail.

File: tests/basic/storage_pool_test.go [hunk 11]
```diff
@@ -10005,3 +10006,341 @@ func exitPoolMaintenance(poolId string) {
...
+var _ = Describe("{PoolDeleteNegative}", func() {
+
+	/*
+				migrated from px-test: PoolDeleteNegative
+		 		1. Delete pool with invalid pool IDs
+				2. Delete pool with volumes on it
+				3. Delete pool after it was added to volume through ha-incr
+	*/
+
+	deletePoolAndValidateFaliure := func(poolID string, selectedNode *node.Node, errRegExp *regexp.Regexp) {
```
Comment (lsrinivas-pure) [inline tests/basic/storage_pool_test.go:10210]: have this inside it. this will give panic if it is not in It

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: drivers/scheduler/k8s/specs/fio-storagepool/fio-config-map.yaml (hunk 8 above)
Comment (dgoel-px) [reply drivers/scheduler/k8s/specs/fio-storagepool/fio-config-map.yaml:8]: mostly to simplify what the spec does, which is create a fio vol and write to it.

File: tests/basic/storage_pool_test.go (hunk 11 above)
Comment (dgoel-px) [reply tests/basic/storage_pool_test.go:10210]: I didn't understand the comment.

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

**Pull request merged**